ckan/lib/webassets_tools.py

```diff
@@ -1,172 +1,244 @@
+        bundle.contents = [
+            os.path.join(path, item)
+            for item in bundle.contents
+        ]
+        log.debug("Register asset %s", name)
         env.register(name, bundle)
 
-    env.append_path(path)
-
 
 def webassets_init() -> None:
+    """Initialize fresh Webassets environment
+    """
     global env
 
     static_path = get_webassets_path()
 
-    public = config.get(u'ckan.base_public_folder')
-
-    public_folder = os.path.abspath(os.path.join(
-        os.path.dirname(__file__), u'..', public))
-
-    base_path = os.path.join(public_folder, u'base')
-
     env = Environment()
     env.directory = static_path
-    env.debug = config.get(u'debug')
-    env.url = u'/webassets/'
+    env.debug = config["debug"] and False
+    env.url = config["ckan.webassets.url"]
 
-    add_public_path(base_path, u'/base/')
 
-    logger.debug(u'Base path {0}'.format(base_path))
-    create_library(u'vendor', os.path.join(
-        base_path, u'vendor'))
+def register_core_assets():
+    """Register CKAN core assets.
 
-    create_library(u'base', os.path.join(base_path, u'javascript'))
+    Call this function after registration of plugin assets. Asset overrides are
+    not alowed, so if plugin tries to replace CKAN core asset, it has to
+    register an asset with the same name before core asset is added. In this
+    case, asset from plugin will have higher precedence and core asset will be
+    ignored.
 
-    create_library(u'datapreview', os.path.join(base_path, u'datapreview'))
+    """
+    public = config["ckan.base_public_folder"]
+    public_folder = os.path.abspath(os.path.join(
+        os.path.dirname(__file__),
+        "..",
+        public,
+    ))
 
-    create_library(u'css', os.path.join(base_path, u'css'))
+    base_path = os.path.join(public_folder, "base")
+    add_public_path(base_path, "/base/")
+    create_library("vendor", os.path.join(base_path, "vendor"))
+    create_library("base", os.path.join(base_path, "javascript"))
+    create_library("css", os.path.join(base_path, "css"))
 
 
-def _make_asset_collection() -> dict[str, Any]:
-    return {u'style': [], u'script': [], u'included': set()}
+def _make_asset_collection() -> AssetCollection:
+    return {"style": [], "script": [], "included": set()}
 
 
 def include_asset(name: str) -> None:
     from ckan.lib.helpers import url_for_static_or_external
-    try:
-        if not g.webassets:
-            raise AttributeError(u'WebAssets not initialized yet')
-    except AttributeError:
-        g.webassets = _make_asset_collection()
-    if name in g.webassets[u'included']:
+
+    if not env:
+        log.debug("Webassets environment is not initialized yet")
         return
 
-    assert env
-    try:
-        bundle: Any = env[name]
-    except KeyError:
-        logger.error(u'Trying to include unknown asset: <{}>'.format(name))
+    if not hasattr(g, "webassets"):
+        g.webassets = _make_asset_collection()
```

Comment:
I would add a log.debug() statement since we are expecting something to have happened but it didn't so we are doing it again. This silent scenarios tends to be difficult to debug.

Reply:
Not exactly. We are initializing collection the first time we visit this function. We are not expecting anything, just making sure that something exists. It's like `dict.setdefault`.